design/open/delta-diff-support.md

```diff
@@ -0,0 +1,179 @@
+# Proposal: Native Delta Lake Diff
+
+## User Story
+
+As a data engineer, I'm working on an isolated branch during development and testing of new pipelines (or modifications to existing ones).
+Since I'm using Delta Lake as a table format, I'd like to understand how changing my code modifies the data I'm testing on.
+
+## Goal
+
+Allow extending the diff functionality beyond the simple object-wise comparison that is currently supported.
+For tables known to be Delta Lake tables, show a logical diff based on the log metadata as exposed by the [Delta History API](https://docs.delta.io/latest/delta-utility.html#retrieve-delta-table-history).
+
+## Non goals
+
+This is still in early stages: this proposal doesn't cover logical merges or other functionality based on the metadata provided by Delta Lake.
+
+
+## Implementation
+
+In order to show a "logical" diff of delta tables, we need the following in place:
+
+1. A way to define a subset of a repository as a Delta Table
+1. Given a delta table, be able to compare 2 versions of it and get a logical diff
+1. UX: Be able to retrieve this from the lakeFS API, SDKs, CLI and UI
+
+### Defining Datasets
+
+Diffing Delta tables requires knowing where a Delta table is located within a repository.
+A single repository can potentially hold any number of Delta tables at different prefixes.
+
+To support this, we will add a new primitive to lakeFS: `Dataset`.
```

Comment:
> I would much rather implement all of these as regular objects
I don't think this proposal contradicts that. How Datasets are implemented under the hood are not covered in this proposal - it can very well be defined using metadata objects with the Datasets tab in the UI being a shiny wrapper on top of that.

> we need to develop a new resource type in lakeFS to advance with tables
Again I don't think that we do. We discussed this internally (@Jonathan-Rosenberg @talSofer and I) so apologies it not covered here - it should be.

We talked about having the first step be a UI element where users can explicitly specify a path to their Delta table to get a "logical" diff, on request. We decided that Datasets are a quality-of-life improvement that could always be added later after we prove need and viability.

> Datasets are versioned and have staging.
I agree with this - if you'll notice, diffing a dataset has a seperate API from the existing diff. The only difference is that the existing diff API will return a "dataset" type instead of "common_prefix" in case the path represents a dataset.
This means that using the existing diff within a dataset will still retain existing behavior: an object-wise diff of the path passed to it regardless of its contents.